OBSDOCS-556: Updating logging alerts docs (#65341)
openshift/openshift-docs: modules/es-cluster-health-is-red.adoc

// Module included in the following assemblies:
//
// * logging/logging_alerts/troubleshooting-logging-alerts.adoc

:_content-type: PROCEDURE

> [id="es-cluster-health-is-red_{context}"]

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696947801776169?thread_ts=1696947268.418629&cid=C02321AKYTT - these need to be lowercase as part of the docs repo guidelines, but once this PR is merged the links shouldn't change again. Let's follow up in Slack to make sure the links in the code get updated properly once these docs are live, since apparently they're broken at the moment anyway.

= Elasticsearch cluster health status is red

At least one primary shard and its replicas are not allocated to a node. Use the following procedure to troubleshoot this alert.

.Procedure

. Check the Elasticsearch cluster health and verify that the cluster `status` is red by running the following command:
+
[source,terminal]
----
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- health
----

. List the nodes that have joined the cluster by running the following command:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=_cat/nodes?v

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="_cat/nodes?v"
```

Author: https://redhat-internal.slack.com/archives/C02321AKYTT/p1696861449261119?thread_ts=1696524477.252959&cid=C02321AKYTT

----

. List the Elasticsearch pods and compare them with the nodes in the command output from the previous step, by running the following command:
+
[source,terminal]
----

> $ oc -n openshift-logging get pods -l component=elasticsearch

Reviewer 2: Maybe something like this (in the beginning of the document, maybe even in the `troubleshooing-logging.adoc`):

---

Most of the commands listed on this page reference an Elasticsearch pod to be executed on and use a shell variable for this (`$ES_POD_NAME`). If you want to copy the commands verbatim out of this page you need to set this variable to a value valid for your Elasticsearch cluster:

```shell
# List the available Elasticsearch pods
$ oc -n openshift-logging get pods -l component=elasticsearch
# Pick one of the pods listed and set the variable
$ export ES_POD_NAME=<elasticsearch-pod-name>
```

Now you can use `$ES_POD_NAME` in the subsequent commands.

---

And then replacing all the mentions of `<elasticsearch-pod-name>` with the variable in our examples.

----

. If some of the Elasticsearch nodes have not joined the cluster, perform the following steps.

> .. Confirm that Elasticsearch has an elected control plane node by running the following command and observing the output:

Reviewer: ```suggestion
.. Confirm that Elasticsearch has an elected master node by running the following command and observing the output:
```

Author: Per https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#using-conscious-language we're supposed to use "control plane node" now.

Author: This is not specific to Kubernetes, it's something that's implemented in many products at Red Hat and is a requirement AFAIK for docs in the OpenShift repo. Since this was preexisting content, I am reluctant to change it to "master", especially because I think it would be pushed back on from the core OCP team.

Author: @kalexand-rh any thoughts on this, is `master` permitted in cases like this?

Reviewer 3: I am not very familiar with Elasticsearch, but from the way it's written, it certainly sounds like it's a node joining an OpenShift cluster, in which case "master" should not be used. If it's an Elasticsearch node joining an Elasticsearch cluster, I suggest specifying "an Elasticsearch cluster" and trying to rewrite the sentence to avoid the term "master." If it's truly the only way to describe it, I think it's reluctantly allowed - we're cleaning up terminology, but I think that clarity is more important. (In other collections, I see "master" used with respect to Apache master nodes.) @bergerhoffer, is there a different way to approach this that we should consider?

Reviewer 4: @kalexand-rh @abrennan89 I am not very familiar with Elasticsearch either.

In general, we should be avoiding the use of "master" if at all possible. The only real cases I'm aware of where we need to keep using it is if it's part of the code/config option (e.g. the `master` role), or is a part of some other non-RH technology/component that uses "master" as part of its terminology.

I can see in looking in the Elasticsearch documentation that they do seem to use the term "master node" throughout. I don't know enough about Elasticsearch to know if they're talking about their own node somewhere outside of the OCP cluster (in which case I'd say that we'd need to keep that term "master node") or if we really are talking about an OCP node in our OCP cluster (in which case I would go with "control plane node").

Another alternative would be to try to rewrite to avoid "master node" looking like we're talking about an OCP control plane node (if we aren't). Maybe by putting Elasticsearch in front of "node" or talking about the node with the `master` role. Not sure what would keep the accuracy here though.

Sorry to not be more helpful. But at the end of the day, we try to avoid it if we can write it another way, but if that really is the term that needs to be used, then we can use it.

Reviewer: A collection of Elasticsearch instances (nodes) is called an Elasticsearch cluster, this is regardless where it is installed, it could be locally on a single machine or on the cloud.

The term "master node" is used throughout ES to indicate an _Elasticsearch_ (not OCP) node that has permissions to create and delete indices. [[ref]](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/modules-node.html#master-node). Maybe it is worth it to clarify this at the beginning of the section.

Reviewer 2: The purpose of this document is to help customers debug issues with Elasticsearch. While I understand why the effort to use more inclusive language exists, I think in this case the usefulness of using the same vocabulary as the product you're debugging outweighs the language concern.

For example, this is the output of the `_cat/nodes` query above:

```
$ oc exec -n openshift-logging -c elasticsearch elasticsearch-cdm-vvnyug3j-1-577b77ccbf-l7vbg -- es_util --query=_cat/nodes?v
ip           heap.percent ram.percent cpu load_1m load_5m load_15m node.role master name
10.128.0.182           50          96   2    5.85    3.51     2.66 mdi       *      elasticsearch-cdm-vvnyug3j-3
10.128.0.181           48          96   2    5.85    3.51     2.66 mdi       -      elasticsearch-cdm-vvnyug3j-2
10.128.0.180           61          93   2    5.85    3.51     2.66 mdi       -      elasticsearch-cdm-vvnyug3j-1
```

The column the user should look at is called `master` and this also carries forward to the endpoint (`_cat/master`) used below. The words "control plane node" do not have any meaning in the context of Elasticsearch.

Author: That makes sense @xperimental , I will update it to master node

+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=_cat/master?v

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="_cat/master?v"
```

Author: https://redhat-internal.slack.com/archives/C02321AKYTT/p1696861449261119?thread_ts=1696524477.252959&cid=C02321AKYTT

----

> .. Review the pod logs of the elected control plane node for issues by running the following command and observing the output:

Reviewer: ```suggestion
.. Review the pod logs of the elected master node for issues by running the following command and observing the output:
```

Author: Per https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#using-conscious-language we're supposed to use "control plane node" now.

+
[source,terminal]
----
$ oc logs <elasticsearch_master_pod_name> -c elasticsearch -n openshift-logging
----

.. Review the logs of nodes that have not joined the cluster for issues by running the following command and observing the output:
+
[source,terminal]
----
$ oc logs <elasticsearch_node_name> -c elasticsearch -n openshift-logging
----

. If all the nodes have joined the cluster, check if the cluster is in the process of recovering by running the following command and observing the output:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=_cat/recovery?active_only=true

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="_cat/recovery?active_only=true"
```

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696866637850469?thread_ts=1696524477.252959&cid=C02321AKYTT we will leave this as is

----
+
If there is no command output, the recovery process might be delayed or stalled by pending tasks.

. Check if there are pending tasks by running the following command and observing the output:
+
[source,terminal]
----
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- health |grep  number_of_pending_tasks
----

. If there are pending tasks, monitor their status. If their status changes and indicates that the cluster is recovering, continue waiting. The recovery time varies according to the size of the cluster and other factors. Otherwise, if the status of the pending tasks does not change, this indicates that the recovery has stalled.

. If it seems like the recovery has stalled, check if the `cluster.routing.allocation.enable` value is set to `none`, by running the following command and observing the output:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=_cluster/settings?pretty

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="_cluster/settings?pretty"
```

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696866637850469?thread_ts=1696524477.252959&cid=C02321AKYTT we will leave this as is

----

. If the `cluster.routing.allocation.enable` value is set to `none`, set it to `all`, by running the following command:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=_cluster/settings?pretty -X PUT -d '{"persistent": {"cluster.routing.allocation.enable":"all"}}'

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="_cluster/settings?pretty" -X PUT -d '{"persistent": {"cluster.routing.allocation.enable":"all"}}'
```

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696866637850469?thread_ts=1696524477.252959&cid=C02321AKYTT we will leave this as is

----

. Check if any indices are still red by running the following command and observing the output:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=_cat/indices?v

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="_cat/indices?v"
```

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696866637850469?thread_ts=1696524477.252959&cid=C02321AKYTT we will leave this as is

----

. If any indices are still red, try to clear them by performing the following steps.

.. Clear the cache by running the following command:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=<elasticsearch_index_name>/_cache/clear?pretty

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="<elasticsearch_index_name>/_cache/clear?pretty"
```

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696866637850469?thread_ts=1696524477.252959&cid=C02321AKYTT we will leave this as is

----

.. Increase the max allocation retries by running the following command:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=<elasticsearch_index_name>/_settings?pretty -X PUT -d '{"index.allocation.max_retries":10}'

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="<elasticsearch_index_name>/_settings?pretty -X PUT -d '{"index.allocation.max_retries":10}'"
```

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696866637850469?thread_ts=1696524477.252959&cid=C02321AKYTT we will leave this as is

----

.. Delete all the scroll items by running the following command:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=_search/scroll/_all -X DELETE

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="_search/scroll/_all -X DELETE"
```

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696866637850469?thread_ts=1696524477.252959&cid=C02321AKYTT we will leave this as is

----

.. Increase the timeout by running the following command:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=<elasticsearch_index_name>/_settings?pretty -X PUT -d '{"index.unassigned.node_left.delayed_timeout":"10m"}'

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="<elasticsearch_index_name>/_settings?pretty" -X PUT -d '{"index.unassigned.node_left.delayed_timeout":"10m"}'
```

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696866637850469?thread_ts=1696524477.252959&cid=C02321AKYTT we will leave this as is

----

. If the preceding steps do not clear the red indices, delete the indices individually.

.. Identify the red index name by running the following command:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=_cat/indices?v

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="_cat/indices?v"
```

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696866637850469?thread_ts=1696524477.252959&cid=C02321AKYTT we will leave this as is

----

.. Delete the red index by running the following command:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=<elasticsearch_red_index_name> -X DELETE

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="<elasticsearch_red_index_name> -X DELETE"
```

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696866637850469?thread_ts=1696524477.252959&cid=C02321AKYTT we will leave this as is

----

. If there are no red indices and the cluster status is red, check for a continuous heavy processing load on a data node.

.. Check if the Elasticsearch JVM Heap usage is high by running the following command:
+
[source,terminal]
----

> $ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query=_nodes/stats?pretty

Reviewer: ```suggestion
$ oc exec -n openshift-logging -c elasticsearch <elasticsearch_pod_name> -- es_util --query="_nodes/stats?pretty"
```

Author: Per https://redhat-internal.slack.com/archives/C02321AKYTT/p1696866637850469?thread_ts=1696524477.252959&cid=C02321AKYTT we will leave this as is

----
+
In the command output, review the `node_name.jvm.mem.heap_used_percent` field to determine the JVM Heap usage.

> .. Check for high CPU utilization.

Author: Maybe, is there more information that you want to provide?

> .. Check for high CPU utilization.

Reviewer 2: Maybe linking to [this page](https://docs.openshift.com/container-platform/4.13/monitoring/reviewing-monitoring-dashboards.html#reviewing-monitoring-dashboards-developer_reviewing-monitoring-dashboards) in the monitoring section is a good fit?

> .. Check for high CPU utilization.

Author: Does this actually apply for ES? I.e. can you actually view CPU usage for it in the OpenShift console, or would it be in Kibana?

> .. Check for high CPU utilization.

Reviewer 2: Yes, from the POV of OpenShift, the Elasticsearch pods are a workload like any other, just that the access to the special `openshift-logging` namespace is usually restricted to a set of admins.

> .. Check for high CPU utilization.

Author: We can't put a link here since it's a module and we'd need to xref to this other file, but I can mention here that they can check this in the web console.

// Module included in the following assemblies:
//
// * logging/logging_alerts/troubleshooting-logging-alerts.adoc

:_content-type: PROCEDURE
[id="es-cluster-health-is-red_{context}"]
= Elasticsearch cluster health status is red

At least one primary shard and its replicas are not allocated to a node. Use the following procedure to troubleshoot this alert.

include::snippets/es-pod-var-logging.adoc[]

.Procedure

. Check the Elasticsearch cluster health and verify that the cluster `status` is red by running the following command:
+
[source,terminal]
----
$ oc exec -n openshift-logging -c elasticsearch $ES_POD_NAME -- health
----

. List the nodes that have joined the cluster by running the following command:
+
[source,terminal]
----
$ oc exec -n openshift-logging -c elasticsearch $ES_POD_NAME -- es_util --query=_cat/nodes?v
----

. List the Elasticsearch pods and compare them with the nodes in the command output from the previous step, by running the following command:
+
[source,terminal]
----
$ oc -n openshift-logging get pods -l component=elasticsearch
----

. If some of the Elasticsearch nodes have not joined the cluster, perform the following steps.

> .. Confirm that Elasticsearch has an elected master node by running the following command and observing the output:

Author: Unfortunately not, I have already discussed it earlier in this PR with Andrea and Kathryn and we decided its OK to leave as is for now.
